api/src/main/java/org/openmrs/api/impl/ConceptServiceImpl.java

```diff
@@ -1982,6 +1984,72 @@ public boolean hasAnyConceptAttribute(ConceptAttributeType conceptAttributeType)
 		return dao.getConceptAttributeCount(conceptAttributeType) > 0;
 	}
 
+	/**
+	 * @see org.openmrs.api.ConceptService#getConceptByNameOrIdOrMap(String conceptRef
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public Concept getConceptByNameOrIdOrMap(String conceptRef) throws APIException {
+		if (conceptRef == null) {
+			return null;
+		}
+		Concept cpt = null;
+		//check if input is a valid Uuid
+		if (!(conceptRef.length() < 36 || conceptRef.length() > 38 || conceptRef.contains(" ") || conceptRef.contains("."))) {
+			cpt = Context.getConceptService().getConceptByUuid(conceptRef);
+			if (cpt != null)
+				return cpt;
+		}
+		//check if input is map
+		if (conceptRef.indexOf(":") != -1) {
+			int idx = conceptRef.indexOf(":");
```

Comment:
Why are we calling `.indexOf()` twice?

```java
int idx = conceptRef.indexOf(":");
if (idx >= 0 && idx < conceptRef.length - 1) {
    String conceptSource = conceptRef.substring(0, idx);
    String conceptCode = conceptRef.substring(idx + 1);
}
```

Also note the extra safety check to avoid the case where conceptRef is something like `CIEL:`.

Reply:
Ohhh Yes! this is really good!